base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -2651,6 +2710,57 @@ fn convert_to_option_bounds<T: RangeBounds<u64>>(bounds: T) -> (Option<u64>, Opt
     (start, end)
 }
 
+/// Process a batch of blocks for PayRef migration
+fn process_payref_for_height<B: BlockchainBackend>(
+    db: Arc<RwLock<B>>,
```

Comment:
```suggestion
    db: &mut B,
```
This should be a single write lock for this entire tx